executor/split.go

```diff
@@ -244,50 +309,73 @@ func (e *SplitTableRegionExec) Next(ctx context.Context, _ *chunk.Chunk) error {
 	}
 	regionIDs := make([]uint64, 0, len(splitKeys))
 	for _, key := range splitKeys {
+		failpoint.Inject("mockSplitRegionTimeout", func(val failpoint.Value) {
+			if val.(bool) {
+				time.Sleep(time.Second*1 + time.Millisecond*10)
+			}
+		})
+		if isCtxDone(ctxWithTimeout) {
+			break
+		}
 		regionID, err := s.SplitRegion(key, true)
 		if err != nil {
 			logutil.BgLogger().Warn("split table region failed",
 				zap.String("table", e.tableInfo.Name.L),
 				zap.Error(err))
 			continue
 		}
+		if regionID == 0 {
+			continue
+		}
 		regionIDs = append(regionIDs, regionID)
 
-		failpoint.Inject("mockSplitRegionTimeout", func(val failpoint.Value) {
-			if val.(bool) {
-				time.Sleep(time.Second * 1)
-			}
-		})
-
-		if isCtxDone(ctxWithTimeout) {
-			return errors.Errorf("split region timeout(%v)", e.ctx.GetSessionVars().GetSplitRegionTimeout())
-		}
 	}
+	e.splitRegions = len(regionIDs)
 	if !e.ctx.GetSessionVars().WaitSplitRegionFinish {
 		return nil
 	}
-	for _, regionID := range regionIDs {
-		err := s.WaitScatterRegionFinish(regionID)
-		if err != nil {
-			logutil.BgLogger().Warn("wait scatter region failed",
-				zap.Uint64("regionID", regionID),
-				zap.String("table", e.tableInfo.Name.L),
-				zap.Error(err))
-		}
 
+	remainMillisecond := 0
+	finishScatterNum := 0
+	for _, regionID := range regionIDs {
 		failpoint.Inject("mockScatterRegionTimeout", func(val failpoint.Value) {
 			if val.(bool) {
-				time.Sleep(time.Second * 1)
+				time.Sleep(time.Second*1 + time.Millisecond*10)
 			}
 		})
-
 		if isCtxDone(ctxWithTimeout) {
-			return errors.Errorf("wait split region scatter timeout(%v)", e.ctx.GetSessionVars().GetSplitRegionTimeout())
+			// Do not break here for checking remain region scatter finished with a very short back off time.
+			// Imagine this situation, we split region 1,2,3, and timeout on wait region 1 scatter,
+			// but region 2 and region 3 is already scatter finish, then we should return result finish scatter region num 2,
+			// instead of finish scatter region num 0.
+			remainMillisecond = 50
```

Comment:
It's better to define a constant instead of `50`

Reply:
done.